xla/python/py_array.cc

```diff
@@ -629,11 +637,68 @@ StatusOr<std::uintptr_t> PyArray::UnsafeBufferPointer() {
       IfrtHelpers::pjrt_buffer(arr.ifrt_array()));
 }
 
-StatusOr<py::dict> PyArray::CudaArrayInterface() {
-  TF_ASSIGN_OR_RETURN(auto arr, AssertUnsharded("UnsafeBufferPointer"));
+py::dict PyArray::CudaArrayInterface() {
+  auto arr = ValueOrThrow(AssertUnsharded("UnsafeBufferPointer"));
+
+  ifrt::Array* ifrt_array = arr.ifrt_array();
+  std::optional<Shape>& scratch = arr.GetStorage().dynamic_shape;
+  auto* pjrt_buffer = IfrtHelpers::pjrt_buffer(ifrt_array);
+  if (pjrt_buffer->client()->platform_id() != CudaId()) {
+    throw py::attribute_error(
+        "__cuda_array_interface__ is only defined for NVidia GPU buffers.");
+  }
+  if (pjrt_buffer->IsTuple()) {
+    throw py::attribute_error(
+        "__cuda_array_interface__ is only defined for array buffers.");
+  }
+
+  switch (pjrt_buffer->element_type()) {
+    case PrimitiveType::PRED:
+    case PrimitiveType::S8:
+    case PrimitiveType::S16:
+    case PrimitiveType::S32:
+    case PrimitiveType::S64:
+    case PrimitiveType::U8:
+    case PrimitiveType::U16:
+    case PrimitiveType::U32:
+    case PrimitiveType::U64:
+    case PrimitiveType::F16:
+    case PrimitiveType::F32:
+    case PrimitiveType::F64:
+    case PrimitiveType::C64:
+    case PrimitiveType::C128:
+      break;
+
+    default:
+      throw py::attribute_error(absl::StrFormat(
+          "__cuda_array_interface__ is not supported for %s buffers.",
+          PrimitiveType_Name(pjrt_buffer->element_type())));
+  }
```

Comment:
This switch block seems to be equivalent to
```c++
if (pjrt_buffer->element_type() == PrimitiveType::BF16) {
  throw ...
}
```
While explicit is better as you have it, less LOC is also good :) Feel free to ignore this comment.

Reply:
No, because there are other types that we're also (implicitly) forbidding here: https://github.com/openxla/xla/blob/511d186d618989f636ba60357a13b2e82f0f5f66/xla/xla_data.proto#L27

e.g., S4/U4, and the various F8 types (e.g., F8E5M2).

I admit I'm not testing all of those in the JAX unit test...

Reply:
Yes, makes sense. Thanks for the explanation!